virttest/libvirt_vm.py

```diff
@@ -281,6 +281,97 @@ def is_autostart(self):
         except process.CmdError:
             return False
 
+    def set_state(self, state, shutoff_method="destroy", debug=False):
+        """
+        Set VM state to "running", "os-booted, "paused", "pmsuspended" or "shutoff"
+
+        param state, the target state of VM
+        param shutoff_method: The method to convert VM state to shutoff, it
+        could be "destroy" or "shutdown"
+        param debug: Enable debug log in logging
+        """
+        target_states = ["running", "os-booted", "paused", "pmsuspended", "shutoff"]
+        abnormal_states = ["no state", "idle", "crashed"]
+        cur_state = self.state()
+        if self.wait_for_login():
+            cur_state = "os-booted"
+
+        if cur_state == "in shutdown":
+            time.sleep(30)
+            if self.state() == "in shutdown":
+                logging.error("VM %s still keep in 'in shutdown' state after 30s waiting",
+                              self.name)
+
+        if cur_state in abnormal_states:
+            logging.error("The VM %s's state %s is abnormal", self.name, cur_state)
+            return
+
+        if states not in target_states:
+            logging.error("Unable to set VM %s's state to %s", self.name, state)
+            return
+
+        if cur_state == state or (cur_state == "os-booted" and state == "running"):
+            logging.info("Current VM %s's state %s is the same with the target state",
+                         self.name, cur_state)
+            return
+
+        if (cur_state == "running" and state == "os-booted"):
+            logging.error("Unable to set the state of VM %s from %s to %s",
+                          self.name, cur_state, state)
+            return
+
+        if cur_state in ["os-booted", "running"]:
+            if state == "paused":
+                virsh.suspend(self.name, uri=self.connect_uri, debug=debug)
```

Comment:
Why don't you use the existing `virsh.pause()`, `virsh.resume()` etc.? Wouldn't this remove code duplication `(self.name, uri=self.connect_uri, debug=debug)`?

Reply:
> Why don't you use the existing `virsh.pause()`, `virsh.resume()` etc.? Wouldn't this remove code duplication 
`(self.name, uri=self.connect_uri, debug=debug)`?

Not such method `virsh.pause()` but `vm.pause()`. `vm.pause()` seems better because of the error handling. I will use it.